universal-application-tool-0.0.1/app/services/applicant/ApplicantServiceImpl.java

```diff
@@ -179,6 +181,104 @@ private void stageUpdates(
       throws UnsupportedScalarTypeException, PathNotInBlockException {
+    if (block.isEnumerator()) {
+      stageEnumeratorUpdates(applicantData, block, updateMetadata, updates);
+    } else {
+      stageNormalUpdates(applicantData, block, updateMetadata, updates);
+    }
+  }
+
+  /**
+   * Stage updates for an enumerator.
+   *
+   * @throws PathNotInBlockException for updates that aren't {@link Scalar#ENTITY_NAME}, or {@link
+   *     Scalar#DELETE_ENTITY}.
+   */
+  private void stageEnumeratorUpdates(
+      ApplicantData applicantData,
+      Block block,
+      UpdateMetadata updateMetadata,
+      ImmutableSet<Update> updates)
+      throws PathNotInBlockException {
+    ImmutableSet<Update> addsAndChanges = validateEnumeratorAddsAndChanges(block, updates);
+    ImmutableSet<Update> deletes =
+        updates.stream()
+            .filter(
+                update ->
+                    update
+                        .path()
+                        .withoutArrayReference()
+                        .equals(Path.empty().join(Scalar.DELETE_ENTITY)))
+            .collect(ImmutableSet.toImmutableSet());
+
+    // If there are more updates than there are adds/changes and deletes, throw
+    Set<Update> unknownUpdates = Sets.difference(updates, Sets.union(addsAndChanges, deletes));
+    if (!unknownUpdates.isEmpty()) {
+      throw new PathNotInBlockException(block.getId(), unknownUpdates.iterator().next().path());
+    }
+
+    for (Update update : addsAndChanges) {
+      applicantData.putString(update.path().join(Scalar.ENTITY_NAME), update.value());
+      writeMetadataForPath(update.path(), applicantData, updateMetadata);
+    }
+
+    ImmutableList<Integer> deleteIndices =
+        deletes.stream()
+            .map(Update::value)
+            .map(Integer::valueOf)
+            .collect(ImmutableList.toImmutableList());
+    applicantData.deleteRepeatedEntities(
+        block.getEnumeratorQuestion().getContextualizedPath(), deleteIndices);
+  }
+
+  /**
+   * Validate that the updates to add or change enumerated entity names have the correct paths with
+   * the right indices.
+   */
+  private ImmutableSet<Update> validateEnumeratorAddsAndChanges(
+      Block block, ImmutableSet<Update> updates) {
+    ImmutableSet<Update> entityUpdates =
+        updates.stream()
+            .filter(
+                update ->
+                    update
+                        .path()
+                        .withoutArrayReference()
+                        .equals(
+                            block
+                                .getEnumeratorQuestion()
+                                .getContextualizedPath()
+                                .withoutArrayReference()))
+            .collect(ImmutableSet.toImmutableSet());
+
+    // Early return if it is empty.
+    if (entityUpdates.isEmpty()) {
+      return entityUpdates;
+    }
+
+    // Check that the entity updates have unique and consecutive indices
```

Comment:
question: why do they need consecutive indices? What if I have 5 children and need to update only the names for #2 and #5?

Reply:
The form will submit all the entity names, so we expect consecutive indices from `0` to `len-1`. Is there a better way to do this?

Reply:
Added a comment to clarify this.